packages/utils/src/utils/agent.utils.ts

```diff
@@ -42,3 +42,102 @@ export const createAgent = async ({
     shouldFetchRootKey: fetchRootKey,
   });
 };
+
+export interface AgentManagerConfig {
+  fetchRootKey: boolean;
+  host: string;
+}
+
+/**
+ * AgentManager class manages HttpAgent instances for different identities.
+ *
+ * It caches agents by identity to optimise resource usage and avoid unnecessary agent creation.
+ * Provides functionality to create new agents, retrieve cached agents, and clear the cache when needed.
+ */
+export class AgentManager {
+  private agents: Record<string, HttpAgent> | undefined = undefined;
+
+  private constructor(private readonly config: AgentManagerConfig) {
+    this.config = config;
+  }
+
+  /**
+   * Static factory method to create a new AgentManager instance.
+   *
+   * This method serves as an alternative to directly using the private constructor,
+   * making it more convenient to create instances of `AgentManager` using a simple and clear method.
+   *
+   * @param {AgentManagerConfig} config - Configuration options for the AgentManager instance.
+   * @param {boolean} config.fetchRootKey - Whether to fetch the root key for certificate validation.
+   * @param {string} config.host - The host to connect to.
+   * @returns {AgentManager} A new instance of `AgentManager`.
+   */
+  public static create(config: AgentManagerConfig): AgentManager {
+    return new AgentManager(config);
+  }
+
+  /**
+   * Get or create an HTTP agent for a given identity.
+   *
+   * If the agent for the specified identity has been created and cached, it is retrieved from the cache.
+   * If no agent exists for the identity, a new one is created, cached, and then returned.
+   *
+   * @param {Identity} identity - The identity to be used to create the agent.
+   * @returns {Promise<HttpAgent>} The HttpAgent associated with the given identity.
+   */
+  public async getAgent({
+    identity,
+  }: {
+    identity: Identity;
+  }): Promise<HttpAgent> {
+    const key = identity.getPrincipal().toText();
+
+    if (isNullish(this.agents) || isNullish(this.agents[key])) {
+      const agent = await this.createAgent({ identity });
+
+      this.agents = {
+        ...(this.agents ?? {}),
+        [key]: agent,
+      };
+
+      return agent;
+    }
+
+    return this.agents[key];
+  }
+
+  /**
+   * Create a new HTTP agent for a given identity.
+   *
+   * This method does not check the cache. It always creates a new agent.
+   * The agent is configured with the specified identity, fetchRootKey option, and the host.
+   *
+   * @param {Identity} identity - The identity to use for the agent.
+   * @param {boolean} [verifyQuerySignatures=true] - Whether to verify query signatures for query calls.
+   * @returns {Promise<HttpAgent>} The newly created HttpAgent instance.
+   */
+  public createAgent({
```

Comment:
This function do not need to be public and can actually be fully removed.

The function it contains / calls, `createAgent`, can be directly called on line 96 instead.

Reply:
yes, right! I put it with the same logic there, just in case, but it makes sense